src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -128,6 +128,14 @@ public static async ValueTask<QuicConnection> ConnectQuicAsync(HttpRequestMessag
                     RemoteEndPoint = endPoint,
                     ClientAuthenticationOptions = clientAuthenticationOptions
                 }, cancellationToken).ConfigureAwait(false);
+
+                if (quicConnection.NegotiatedApplicationProtocol != SslApplicationProtocol.Http3)
+                {
+                    await quicConnection.DisposeAsync().ConfigureAwait(false);
+                    throw new HttpRequestException(HttpRequestError.ConnectionError, "QUIC connected but no HTTP/3 indicated via ALPN.", null, RequestRetryType.RetryOnConnectionFailure);
+                }
```

Comment:
Nit: this check feels like it should be in the http-3-specific logic.
Also, I don't think the `RequestRetryType.RetryOnConnectionFailure` would be honored here given that the exception gets wrapped again.

Reply:
Fixed, moved back to the H3 pool